File: frontend/lib/models/audit_log.dart
import 'audit_entity_type.dart';
import 'audit_operation.dart';

class AuditLog {
  final String id;
  final AuditOperation operation;
  final AuditEntityType entityType;
  final String entityId;
  final String performedBy;
  final String groupId;
  final DateTime performedAt;
  final String? changeDescription;

  AuditLog({
    required this.id,
    required this.operation,
    required this.entityType,
    required this.entityId,
    required this.performedBy,
    required this.groupId,
    required this.performedAt,
    this.changeDescription,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    return AuditLog(
      id: json['id'] as String,
      operation: AuditOperation.fromString(json['operation'] as String),
      entityType: AuditEntityType.fromString(json['entityType'] as String),
      entityId: json['entityId'] as String,
      performedBy: json['performedBy'] as String,
      groupId: json['groupId'] as String,
      performedAt: DateTime.parse(json['performedAt'] as String),
      changeDescription: json['changeDescription'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'operation': operation.name,
      'entityType': entityType.name,
      'entityId': entityId,
      'performedBy': performedBy,
      'groupId': groupId,
      'performedAt': performedAt.toIso8601String(),
      'changeDescription': changeDescription,
    };
  }
}

File: backend/src/main/java/pl/btsoftware/backend/audit/domain/AuditLog.java
package pl.btsoftware.backend.audit.domain;

import static java.time.OffsetDateTime.now;
import static java.time.ZoneOffset.UTC;

import java.time.OffsetDateTime;
import org.jetbrains.annotations.Nullable;
import pl.btsoftware.backend.users.domain.GroupId;
import pl.btsoftware.backend.users.domain.UserId;

public record AuditLog(
        AuditLogId id,
        AuditOperation operation,
        AuditEntityType entityType,
        EntityId entityId,
        UserId performedBy,
        GroupId groupId,
        OffsetDateTime performedAt,
        @Nullable String changeDescription) {
    public static AuditLog create(
            AuditOperation operation,
            AuditEntityType entityType,
            EntityId entityId,
            UserId performedBy,
            GroupId groupId,
            String changeDescription) {
        return new AuditLog(
                AuditLogId.generate(),
                operation,
                entityType,
                entityId,
                performedBy,
                groupId,
                now(UTC),
                changeDescription);
    }
}
